### ISSUE TITLE
Windows arm64 support (aarch64-pc-windows-msvc)

### ISSUE THREAD

**Issue by Alovchin91** [briansmith/ring#1167]
Hi,

I'm trying to compile reqwest with rustls support on Windows arm64 and I'm getting the following error:

```
error: failed to run custom build command for `ring v0.16.19`

Caused by:
  process didn't exit successfully: `C:\Projects\reqwest\target\debug\build\ring-1879f9f353eaed0d\build-script-build` (exit code: 101)
  --- stdout
  OPT_LEVEL = Some("0")
  TARGET = Some("aarch64-pc-windows-msvc")
  HOST = Some("aarch64-pc-windows-msvc")
  CC_aarch64-pc-windows-msvc = None
  CC_aarch64_pc_windows_msvc = None
  HOST_CC = None
  CC = None
  CFLAGS_aarch64-pc-windows-msvc = None
  CFLAGS_aarch64_pc_windows_msvc = None
  HOST_CFLAGS = None
  CFLAGS = None
  CRATE_CC_NO_DEFAULTS = None
  CARGO_CFG_TARGET_FEATURE = None
  DEBUG = Some("true")
  aes_nohw.c
  C:\Users\alovchin\.cargo\registry\src\github.com-1ecc6299db9ec823\ring-0.16.19\include\GFp/base.h(97): fatal error C1189: #error:  "Unknown target CPU"

  --- stderr
  running "C:\\Program Files (x86)\\Microsoft Visual Studio\\2019\\BuildTools\\VC\\Tools\\MSVC\\14.28.29333\\bin\\HostX86\\ARM64\\cl.exe" "-nologo" "-MD" "-Z7" "-Brepro" "-I" "include" "-W4" "/GS" "/Gy" "/EHsc" "/GR-" "/Zc:wchar_t" "/Zc:forScope" "/Zc:inline" "/Zc:rvalueCast" "/sdl" "/Wall" "/wd4127" "/wd4464" "/wd4514" "/wd4710" "/wd4711" "/wd4820" "/wd5045" "/Od" "/RTCsu" "-DNDEBUG" "-c" "/FoC:\\Projects\\reqwest\\target\\debug\\build\\ring-b786d3340d88e8a0\\out\\aes_nohw.obj" "crypto/fipsmodule/aes/aes_nohw.c"
  thread 'main' panicked at 'execution failed', C:\Users\alovchin\.cargo\registry\src\github.com-1ecc6299db9ec823\ring-0.16.19\build.rs:673:9
```

I think it's somewhat related to #960 though in this case I'm working on aarch64-pc-windows-msvc platform which is **not** UWP.

This is currently blocking https://github.com/rust-lang/rustup/issues/2612 as rustup is using reqwest's rustls feature.

Please note that unfortunately there is no hosted CI/CD to run tests on Windows arm64 platform yet, so this would be cross-compile only for now.

**Comment by briansmith** [comment]
For Aarch64 Windows targets, we need to review the Windows AArch64 ABI to see if it is the same as Linux. If so then we need to see if PerlAsm works for Aarch64 Windows assembly. If so then we should be golden. If so then presumably we need to use clang and not MSVC unless the MS toolchain includes a GCC-/clang- compatible assembly for AAarch64.

Unless/until we can figure out all of that, we'd need to skip the assembly code completely. This means we'd need to finish merging some of the pending PRs for *ring* to have non-assembly fallbacks for all assembly code. I'll work on that.

**Comment by Alovchin91** [comment]
I'm afraid I cannot help much here as I'm not in any way proficient in assembly code 😬

But maybe this can help: https://docs.microsoft.com/en-us/cpp/build/arm64-windows-abi-conventions

As for the assembler, MSVC seems to include armasm64.exe which is somewhat described here: https://docs.microsoft.com/en-us/cpp/assembler/arm/arm-assembler-reference

The documentation talks about armasm.exe but as far as I understand it, armasm64.exe is very similar. Yet again, I cannot say anything about what dialect of assembly language does it accept (if there is any difference for ARM64).

**Comment by briansmith** [comment]
Good news: BoringSSL recently merged the Windows ARM support, so this might now be much easier: https://boringssl.googlesource.com/boringssl/+/afd5dba756b6266fa99c11af6496b39d826769cd

**Comment by Alovchin91** [comment]
Hi @briansmith Would you please find some time to look into this? Thanks a lot!

**Comment by awakecoding** [comment]
@briansmith @Alovchin91 any update on this? I just tried building some of my Rust code for Windows ARM64, and ring v0.16.20 fails to build for me with the same error as this ticket.

**Comment by Alovchin91** [comment]
@briansmith  I'm also blocked on this with Rustup unfortunately 😞  Do you have any idea of what should be done to enable Windows ARM64 compilation?  Maybe me and @awakecoding could help you somehow?

**Comment by awakecoding** [comment]
@Alovchin91 I've been working on porting 20+ native dependencies to Windows ARM64 for Devolutions, I can definitely help, but I know absolutely nothing about the ring internals and how it builds the part of code that fails. I can take a closer look later this week.

**Comment by awakecoding** [comment]
I inspected the upstream boringssl changes, and some of them appear to have already been ported to the current sources. I decided to take a closer look at the build failure which is caused by the BN_ULLONG type not being defined while being used from mongomery.c:

![image](https://user-images.githubusercontent.com/295841/127526497-34e1c3a1-2d6f-4fbe-a692-9b19c92f81dc.png)

Most of the time build failures result from assumptions that there are just two platforms: 32-bit intel, and 64-bit intel. ARM64 is 64-bit so it goes in the "OPENSSL_64_BIT" ifdef, but it doesn't have the "BORINGSSL_HAS_UINT128" definition. I tried manually forcing it on, but that doesn't look like those types do not exist in MSVC:

![image](https://user-images.githubusercontent.com/295841/127527040-765e4a39-b944-4687-bf28-3abdc30bd65b.png)

Is there a 128-bit integer type in MSVC? If not, how exactly do we disable the usage of BN_ULLONG, or it is mandatory? I tried the dumb thing of defining it to uint64_t but it results in a warning treated as a error due to excessive bit shifting leading to undefined behavior.

Maybe @briansmith can provide some more clarity on this part of the code

**Comment by tommyvct** [comment]
> Is there a 128-bit integer type in MSVC?

As far as I know, the answer is no they don't. But it can be implemented by using 2 64-bit long int like this:  
https://github.com/yuzu-emu/yuzu/blob/master/src/common/common_types.h#L47  
https://github.com/yuzu-emu/yuzu/blob/c5ca8675c84ca73375cf3fe2ade257c8aa5c1239/src/common/uint128.h#L43-L54

**Comment by Alovchin91** [comment]
@awakecoding It should be rather simple to patch `BN_ULLONG`:

```
diff --git a/crypto/fipsmodule/bn/internal.h b/crypto/fipsmodule/bn/internal.h
index c3ba88e23..a2481a666 100644
--- a/crypto/fipsmodule/bn/internal.h
+++ b/crypto/fipsmodule/bn/internal.h
@@ -185,8 +185,13 @@ void bn_mul_mont(BN_ULONG *rp, const BN_ULONG *ap, const BN_ULONG *bp,
 
 static inline void bn_umult_lohi(BN_ULONG *low_out, BN_ULONG *high_out,
                                  BN_ULONG a, BN_ULONG b) {
-#if defined(OPENSSL_X86_64) && defined(_MSC_VER) && !defined(__clang__)
+#if defined(_MSC_VER) && !defined(__clang__)
+#if defined(OPENSSL_X86_64)
   *low_out = _umul128(a, b, high_out);
+#elif defined(OPENSSL_AARCH64)
+  *low_out = a * b;
+  *high_out = __umulh(a, b);
+#endif
 #else
   BN_ULLONG result = (BN_ULLONG)a * b;
   *low_out = (BN_ULONG)result;
```

(courtesy of https://github.com/microsoft/winrtc/blob/master/patches_for_WebRTC_org/m84/4001-Arm64-is-a-thing-and-has-intrinsic-to-mul-two-64bit-.patch)

But then there's `limbs.inl` and x86 intrinsics there, with an alternative being uint128_t...

https://github.com/briansmith/ring/blob/fe67830da183a3254b0a9da179631f1c9631d2f7/crypto/limbs/limbs.inl#L37-L53

**Comment by Alovchin91** [comment]
I have a WIP branch here: https://github.com/briansmith/ring/compare/main...Alovchin91:alovchin91/win-arm64?expand=1

Please note that I have no idea what am I doing so it might be completely wrong 😆 Also I don't seem to be able to build the tests yet:

```
          libring-de44f9620d9cd437.rlib(ring-de44f9620d9cd437.2yt3f83b8gv3ubii.rcgu.o) : error LNK2019: unresolved external symbol ring_core_0_17_0_not_released_yet_ChaCha20_ctr32 referenced in function _ZN4ring4aead6chacha3Key17encrypt_less_safe17h116490da96c8b6e1E
          libring-de44f9620d9cd437.rlib(ring-de44f9620d9cd437.17er0ilt3zov722z.rcgu.o) : error LNK2019: unresolved external symbol ring_core_0_17_0_not_released_yet_aes_hw_set_encrypt_key referenced in function _ZN4ring4aead3aes3Key3new17h48f2a62ee5fd41f4E
          libring-de44f9620d9cd437.rlib(ring-de44f9620d9cd437.17er0ilt3zov722z.rcgu.o) : error LNK2019: unresolved external symbol ring_core_0_17_0_not_released_yet_vpaes_set_encrypt_key referenced in function _ZN4ring4aead3aes3Key3new17h48f2a62ee5fd41f4E
          ...
```

etc. so I don't think you can actually use it in any binary yet.

@briansmith Would you kindly have a look at it and maybe point me in the direction where I could find a solution? 🙂

**Comment by awakecoding** [comment]
@Alovchin91 I made a branch in which I imported your patches, it does build, but the test executable won't link with the same unresolved externals. I have a feeling that maybe the assembly code is not getting correctly built - are the unresolved externals all assembly functions? One of the difficult things with Windows ARM64 porting is its weird assembler program (armasm64.exe). The Windows assembler program for Intel appears to be nasm, and the current configuration you've set for Windows ARM64 appears to imply the sources should contain precompiled object files. Is there a way to confirm if we're getting the object files at all for the missing symbols?

![image](https://user-images.githubusercontent.com/295841/128609101-da6e858e-596e-4398-bafb-cc9dfbbb66a6.png)

**Comment by awakecoding** [comment]
Ok, so I digged a little deeper into the build system to figure out what it does - on Windows / intel it uses nasm (for some odd reason it doesn't seem to pick up nasm.exe from my path, but expects it to be inside the target/tools/windows/nasm directory). I don't know much about perlasm but my guess is it's one of those assembler format conversion tools to help adapt the same assembly code to different assemblers. I have no idea which assembler it calls, but it does produce *something* with the added Windows aarch64 configuration. The interesting part is the ring_core_generated directory that contains both a C header and an asm include file with all the exported symbols - those same missing symbols we don't have. I suspect maybe perlasm doesn't get the right combination of parameters, or it simply doesn't do the right thing, and we either don't end up with the right declarations, or we're not getting the symbols for the right architecture here and they get ignored.

![image](https://user-images.githubusercontent.com/295841/128609733-1855fef9-3cad-4258-a2bf-8e96ffeb239b.png)

![image](https://user-images.githubusercontent.com/295841/128609753-c2b0dd1d-7fed-4171-8dce-294743a8da07.png)

**Comment by awakecoding** [comment]
I think the current build system simply has no support for the Windows ARM64 assembler. If you set the RING_PREGENERATE_ASM environment variable to 1, it'll try pre-generating everything for the assembly files. I modified ASM_TARGETS to contain only Windows ARM64 so I could focus only on that one. The pre-generation fails because it tries calling nasm on the assembly files, which is obviously unsupported. At least now we have something to work with - try and modify the build system to make sure it does call a working assembler program for Windows ARM64 (armasm64.exe or clang.exe).

![image](https://user-images.githubusercontent.com/295841/128612825-153478ed-bfc3-4f5b-ae35-d9a91382e91a.png)

**Comment by awakecoding** [comment]
Ok, armasm64.exe is basically a very obscure tool with a syntax I'm not sure a lot of people know about, so my attempts at calling it failed. However, I did manage to get ring built for Windows ARM64 using clang.exe as the assembler program! I have a bunch of unclean local changes, but this seems to be the way to go:

![image](https://user-images.githubusercontent.com/295841/128613320-b81a6a8f-7ad6-478d-b39c-79ccaf3b2dc3.png)

**Comment by awakecoding** [comment]
@Alovchin91 ok if you want to give it a try, here's here to do it using my branch (https://github.com/awakecoding/ring/tree/windows-arm64):

first, copy nasm.exe to target/tools/windows/nasm/nasm.exe in the ring source tree, that's for intel assembly.
Set RING_PREGENERATE_ASM to 1 ($Env:RING_PREGENERATE_ASM='1')
Make sure clang.exe is in your PATH  (use either the [official clang+llvm build](https://github.com/llvm/llvm-project/releases/tag/llvmorg-12.0.1) or [mine](https://github.com/awakecoding/llvm-prebuilt/releases/tag/v12.0.1))
Build the project (cargo -vv build --target aarch64-pc-windows-msvc)
And then try building the test executables (cargo -vv test --target aarch64-pc-windows-msvc)

You will no longer get the unresolved symbols, but the test code will fail to build properly. If you have issues with pregeneration, just delete the "pregenerated" directory and it should re-generate correctly.

Let me know if you get any further!

**Comment by Alovchin91** [comment]
@awakecoding Looks promising! I'll probably have a look at it tomorrow 🙂

FYI I've managed to build my branch having Perl in the PATH (I've used Git for Windows bash under x64 emulation for this).

**Comment by Alovchin91** [comment]
Okay, I didn't look into your code yet @awakecoding but here's what I've found regarding armasm64.exe.

First of all, it appears that armasm.exe and armasm64.exe are basically the same tool when speaking of MSVC. It has some scarce documentation but it appears to be enough because:

> For the most part, the Microsoft ARM assembler uses the ARM assembly language, which is documented in the ARM Compiler armasm Reference Guide.

https://docs.microsoft.com/en-us/cpp/assembler/arm/arm-assembler-directives?view=msvc-160

Second, it looks like some manual setup is required to use armasm64.exe with tools like Cmake, but it shouldn't be an issue since we're using manually written build.rs anyway.

So it looks like the first step would be to invoke armasm64.exe from build.rs instead of nasm.exe for Windows ARM. The latest version of cc (1.0.69 and up) should make it as easy as calling `cc::windows_registry::find_tool(&target, "armasm64.exe")`.

Here are examples of how .NET team does that (using Cmake; found [here](https://stackoverflow.com/questions/59742466/building-arm64-asm-in-visual-studio-with-cmake)):

https://github.com/dotnet/runtime/blob/f8f63b1fde85119c925313caa475d9936297b463/eng/native/functions.cmake#L173-L207

https://github.com/dotnet/runtime/blob/f8f63b1fde85119c925313caa475d9936297b463/eng/native/configurecompiler.cmake#L611-L626

Regarding the perlasm_format being "win64", I just copied it from the [commit](https://boringssl.googlesource.com/boringssl/+/afd5dba756b6266fa99c11af6496b39d826769cd%5E%21/#F1) mentioned by @briansmith in January 🙂

**Comment by awakecoding** [comment]
@Alovchin91 my original attempt was to call armasm64.exe instead of clang, but even once I got the right command-line parameters, it just wouldn't accept pretty much the entire file with syntax errors and unrecognized stuff, while clang.exe took it instantly like a big boy.

You can give it a shot, just do a search in the visual studio directories to find the executable and then manually add it to your path, and edit the code on my branch to call it instead. Just invoke armasm64.exe -h to see what parameters it takes.

I recently updated my libvpx builds with ARM64 support and it does seem to be calling armasm64.exe - it has a rather simple perl script for adapting the syntax, so I guess it just have been reusing something else I haven't found.

I would suggest going the clang.exe route at least for now to get everything built correctly. Once you go past the missing symbols, it fails to compile the actual test code, it may not require much but it still need to be done.

There is also the possibility of improving support for the Windows assemblers - nasm is a bit tricky because it has to be copied over, but it would be trivial to add a detection function. I'm thinking we could support both clang.exe and armasm64.exe and select the one we want with an environment variable + a default.

**Comment by Alovchin91** [comment]
In fact, I noticed just now that the message of the commit that introduces ASM support for Windows on ARM actually explicitly states that it expects Clang syntax:

> Note these files use GNU assembler syntax (specifically tested with Clang assembler), not armasm.

How didn't I notice that before 🤦‍♂️

It's probably not worth the effort to support both, especially since it's mostly on BoringSSL side of things, but I guess we should leave this decision to @briansmith 🙂

I'll have a look at the test build failures when I get home.

BTW my math in the previous commits is likely terrible and might not work 🙈  Also I'm not so sure about [ignoring the 4133 warning](https://github.com/Alovchin91/ring/commit/7517ee22561151f34c65821c1e05cbc7dd738f99).

**Comment by Alovchin91** [comment]
I've used your code @awakecoding and I've also changed the extension to .asm so that clang.exe does actually get called:

https://github.com/Alovchin91/ring/commit/62cc4afa0eb8bec18f527f2ef2a5e02b7ad34edf

Now I've stumbled upon another issue. Apparently for some reason clang doesn't want to `#include <ring-core/arm_arch.h>` so I get a bunch of errors like the following:

```
  running "clang.exe" "-o" "C:\\Projects\\ring\\target\\debug\\build\\ring-1b847110242e0888\\out\\aesv8-armx-win64.obj" "-I" "include/" "-I" "C:\\Projects\\ring\\target\\debug\\build\\ring-1b847110242e0888\\out\\" "-c" "C:\\Projects\\ring\\target\\debug\\build\\ring-1b847110242e0888\\out\\aesv8-armx-win64.asm"
  C:\Projects\ring\target\debug\build\ring-1b847110242e0888\out\aesv8-armx-win64.asm:37:2: error: unrecognized instruction mnemonic
          AARCH64_VALID_CALL_TARGET
          ^
...
```

In the end I've managed to get past assembly compilation by replacing them all with the respective macro values, see the following diff.

<details>
<summary>the following diff</summary>

```diff
diff --git a/crypto/chacha/asm/chacha-armv4.pl b/crypto/chacha/asm/chacha-armv4.pl
index bb63413f4..5fe677ddd 100755
--- a/crypto/chacha/asm/chacha-armv4.pl
+++ b/crypto/chacha/asm/chacha-armv4.pl
@@ -231,7 +231,7 @@ ChaCha20_ctr32:
 # ifdef	__APPLE__
 	ldr	r4,[r4]
 # endif
-	tst	r4,#ARMV7_NEON
+	tst	r4,#(1 << 0)
 	bne	.LChaCha20_neon
 .Lshort:
 #endif
diff --git a/crypto/chacha/asm/chacha-armv8.pl b/crypto/chacha/asm/chacha-armv8.pl
index f40d66a88..326442fe0 100755
--- a/crypto/chacha/asm/chacha-armv8.pl
+++ b/crypto/chacha/asm/chacha-armv8.pl
@@ -140,7 +140,7 @@ $code.=<<___;
 .type	ChaCha20_ctr32,%function
 .align	5
 ChaCha20_ctr32:
-	AARCH64_VALID_CALL_TARGET
+	// AARCH64_VALID_CALL_TARGET
 	cbz	$len,.Labort
 #if __has_feature(hwaddress_sanitizer) && __clang_major__ >= 10
 	adrp	@x[0],:pg_hi21_nc:OPENSSL_armcap_P
@@ -150,11 +150,11 @@ ChaCha20_ctr32:
 	cmp	$len,#192
 	b.lo	.Lshort
 	ldr	w17,[@x[0],:lo12:OPENSSL_armcap_P]
-	tst	w17,#ARMV7_NEON
+	tst	w17,#(1 << 0)
 	b.ne	ChaCha20_neon
 
 .Lshort:
-	AARCH64_SIGN_LINK_REGISTER
+	// AARCH64_SIGN_LINK_REGISTER
 	stp	x29,x30,[sp,#-96]!
 	add	x29,sp,#0
 
@@ -275,7 +275,7 @@ $code.=<<___;
 	ldp	x25,x26,[x29,#64]
 	ldp	x27,x28,[x29,#80]
 	ldp	x29,x30,[sp],#96
-	AARCH64_VALIDATE_LINK_REGISTER
+	// AARCH64_VALIDATE_LINK_REGISTER
 .Labort:
 	ret
 
@@ -332,7 +332,7 @@ $code.=<<___;
 	ldp	x25,x26,[x29,#64]
 	ldp	x27,x28,[x29,#80]
 	ldp	x29,x30,[sp],#96
-	AARCH64_VALIDATE_LINK_REGISTER
+	// AARCH64_VALIDATE_LINK_REGISTER
 	ret
 .size	ChaCha20_ctr32,.-ChaCha20_ctr32
 ___
@@ -378,7 +378,7 @@ $code.=<<___;
 .type	ChaCha20_neon,%function
 .align	5
 ChaCha20_neon:
-	AARCH64_SIGN_LINK_REGISTER
+	// AARCH64_SIGN_LINK_REGISTER
 	stp	x29,x30,[sp,#-96]!
 	add	x29,sp,#0
 
@@ -578,7 +578,7 @@ $code.=<<___;
 	ldp	x25,x26,[x29,#64]
 	ldp	x27,x28,[x29,#80]
 	ldp	x29,x30,[sp],#96
-	AARCH64_VALIDATE_LINK_REGISTER
+	// AARCH64_VALIDATE_LINK_REGISTER
 	ret
 
 .Ltail_neon:
@@ -688,7 +688,7 @@ $code.=<<___;
 	ldp	x25,x26,[x29,#64]
 	ldp	x27,x28,[x29,#80]
 	ldp	x29,x30,[sp],#96
-	AARCH64_VALIDATE_LINK_REGISTER
+	// AARCH64_VALIDATE_LINK_REGISTER
 	ret
 .size	ChaCha20_neon,.-ChaCha20_neon
 ___
@@ -701,7 +701,7 @@ $code.=<<___;
 .type	ChaCha20_512_neon,%function
 .align	5
 ChaCha20_512_neon:
-	AARCH64_SIGN_LINK_REGISTER
+	// AARCH64_SIGN_LINK_REGISTER
 	stp	x29,x30,[sp,#-96]!
 	add	x29,sp,#0
 
@@ -1121,7 +1121,7 @@ $code.=<<___;
 	ldp	x25,x26,[x29,#64]
 	ldp	x27,x28,[x29,#80]
 	ldp	x29,x30,[sp],#96
-	AARCH64_VALIDATE_LINK_REGISTER
+	// AARCH64_VALIDATE_LINK_REGISTER
 	ret
 .size	ChaCha20_512_neon,.-ChaCha20_512_neon
 ___
diff --git a/crypto/fipsmodule/aes/asm/aesv8-armx.pl b/crypto/fipsmodule/aes/asm/aesv8-armx.pl
index 89cc6c573..03ce35bee 100644
--- a/crypto/fipsmodule/aes/asm/aesv8-armx.pl
+++ b/crypto/fipsmodule/aes/asm/aesv8-armx.pl
@@ -97,7 +97,7 @@ ${prefix}_set_encrypt_key:
 ___
 $code.=<<___	if ($flavour =~ /64/);
 	// Armv8.3-A PAuth: even though x30 is pushed to stack it is not popped later.
-	AARCH64_VALID_CALL_TARGET
+	// AARCH64_VALID_CALL_TARGET
 	stp	x29,x30,[sp,#-16]!
 	add	x29,sp,#0
 ___
@@ -251,7 +251,7 @@ $code.=<<___;
 .type	${prefix}_${dir}crypt,%function
 .align	5
 ${prefix}_${dir}crypt:
-	AARCH64_VALID_CALL_TARGET
+	// AARCH64_VALID_CALL_TARGET
 	ldr	$rounds,[$key,#240]
 	vld1.32	{$rndkey0},[$key],#16
 	vld1.8	{$inout},[$inp]
@@ -304,7 +304,7 @@ ${prefix}_ctr32_encrypt_blocks:
 ___
 $code.=<<___	if ($flavour =~ /64/);
 	// Armv8.3-A PAuth: even though x30 is pushed to stack it is not popped later.
-	AARCH64_VALID_CALL_TARGET
+	// AARCH64_VALID_CALL_TARGET
 	stp		x29,x30,[sp,#-16]!
 	add		x29,sp,#0
 ___
diff --git a/crypto/fipsmodule/aes/asm/vpaes-armv8.pl b/crypto/fipsmodule/aes/asm/vpaes-armv8.pl
index c5f1b49f6..47323a53d 100755
--- a/crypto/fipsmodule/aes/asm/vpaes-armv8.pl
+++ b/crypto/fipsmodule/aes/asm/vpaes-armv8.pl
@@ -239,7 +239,7 @@ _vpaes_encrypt_core:
 .type	vpaes_encrypt,%function
 .align	4
 vpaes_encrypt:
-	AARCH64_SIGN_LINK_REGISTER
+	// AARCH64_SIGN_LINK_REGISTER
 	stp	x29,x30,[sp,#-16]!
 	add	x29,sp,#0
 
@@ -249,7 +249,7 @@ vpaes_encrypt:
 	st1	{v0.16b}, [$out]
 
 	ldp	x29,x30,[sp],#16
-	AARCH64_VALIDATE_LINK_REGISTER
+	// AARCH64_VALIDATE_LINK_REGISTER
 	ret
 .size	vpaes_encrypt,.-vpaes_encrypt
 
@@ -395,7 +395,7 @@ _vpaes_key_preheat:
 .type	_vpaes_schedule_core,%function
 .align	4
 _vpaes_schedule_core:
-	AARCH64_SIGN_LINK_REGISTER
+	// AARCH64_SIGN_LINK_REGISTER
 	stp	x29, x30, [sp,#-16]!
 	add	x29,sp,#0
 
@@ -555,7 +555,7 @@ _vpaes_schedule_core:
 	eor	v6.16b, v6.16b, v6.16b		// vpxor	%xmm6,	%xmm6,	%xmm6
 	eor	v7.16b, v7.16b, v7.16b		// vpxor	%xmm7,	%xmm7,	%xmm7
 	ldp	x29, x30, [sp],#16
-	AARCH64_VALIDATE_LINK_REGISTER
+	// AARCH64_VALIDATE_LINK_REGISTER
 	ret
 .size	_vpaes_schedule_core,.-_vpaes_schedule_core
 
@@ -726,7 +726,7 @@ _vpaes_schedule_mangle:
 .type	vpaes_set_encrypt_key,%function
 .align	4
 vpaes_set_encrypt_key:
-	AARCH64_SIGN_LINK_REGISTER
+	// AARCH64_SIGN_LINK_REGISTER
 	stp	x29,x30,[sp,#-16]!
 	add	x29,sp,#0
 	stp	d8,d9,[sp,#-16]!	// ABI spec says so
@@ -742,7 +742,7 @@ vpaes_set_encrypt_key:
 
 	ldp	d8,d9,[sp],#16
 	ldp	x29,x30,[sp],#16
-	AARCH64_VALIDATE_LINK_REGISTER
+	// AARCH64_VALIDATE_LINK_REGISTER
 	ret
 .size	vpaes_set_encrypt_key,.-vpaes_set_encrypt_key
 ___
@@ -758,7 +758,7 @@ $code.=<<___;
 .type	vpaes_ctr32_encrypt_blocks,%function
 .align	4
 vpaes_ctr32_encrypt_blocks:
-	AARCH64_SIGN_LINK_REGISTER
+	// AARCH64_SIGN_LINK_REGISTER
 	stp	x29,x30,[sp,#-16]!
 	add	x29,sp,#0
 	stp	d8,d9,[sp,#-16]!	// ABI spec says so
@@ -826,7 +826,7 @@ vpaes_ctr32_encrypt_blocks:
 	ldp	d10,d11,[sp],#16
 	ldp	d8,d9,[sp],#16
 	ldp	x29,x30,[sp],#16
-	AARCH64_VALIDATE_LINK_REGISTER
+	// AARCH64_VALIDATE_LINK_REGISTER
 	ret
 .size	vpaes_ctr32_encrypt_blocks,.-vpaes_ctr32_encrypt_blocks
 ___
diff --git a/crypto/fipsmodule/bn/asm/armv4-mont.pl b/crypto/fipsmodule/bn/asm/armv4-mont.pl
index f9af79503..cc07df365 100644
--- a/crypto/fipsmodule/bn/asm/armv4-mont.pl
+++ b/crypto/fipsmodule/bn/asm/armv4-mont.pl
@@ -136,7 +136,7 @@ bn_mul_mont:
 #ifdef	__APPLE__
 	ldr	r0,[r0]
 #endif
-	tst	r0,#ARMV7_NEON		@ NEON available?
+	tst	r0,#(1 << 0)		@ NEON available?
 	ldmia	sp, {r0,r2}
 	beq	.Lialu
 	add	sp,sp,#8
diff --git a/crypto/fipsmodule/bn/asm/armv8-mont.pl b/crypto/fipsmodule/bn/asm/armv8-mont.pl
index f82c495e8..6b4fd7542 100644
--- a/crypto/fipsmodule/bn/asm/armv8-mont.pl
+++ b/crypto/fipsmodule/bn/asm/armv8-mont.pl
@@ -72,7 +72,7 @@ $code.=<<___;
 .type	bn_mul_mont,%function
 .align	5
 bn_mul_mont:
-	AARCH64_SIGN_LINK_REGISTER
+	// AARCH64_SIGN_LINK_REGISTER
 	tst	$num,#7
 	b.eq	__bn_sqr8x_mont
 	tst	$num,#3
@@ -270,7 +270,7 @@ bn_mul_mont:
 	mov	x0,#1
 	ldp	x23,x24,[x29,#48]
 	ldr	x29,[sp],#64
-	AARCH64_VALIDATE_LINK_REGISTER
+	// AARCH64_VALIDATE_LINK_REGISTER
 	ret
 .size	bn_mul_mont,.-bn_mul_mont
 ___
@@ -1047,7 +1047,7 @@ $code.=<<___;
 	ldp	x27,x28,[x29,#80]
 	ldr	x29,[sp],#128
 	// x30 is popped earlier
-	AARCH64_VALIDATE_LINK_REGISTER
+	// AARCH64_VALIDATE_LINK_REGISTER
 	ret
 .size	__bn_sqr8x_mont,.-__bn_sqr8x_mont
 ___
@@ -1508,7 +1508,7 @@ __bn_mul4x_mont:
 	ldp	x27,x28,[x29,#80]
 	ldr	x29,[sp],#128
 	// x30 is popped earlier
-	AARCH64_VALIDATE_LINK_REGISTER
+	// AARCH64_VALIDATE_LINK_REGISTER
 	ret
 .size	__bn_mul4x_mont,.-__bn_mul4x_mont
 ___
diff --git a/crypto/fipsmodule/modes/asm/ghash-neon-armv8.pl b/crypto/fipsmodule/modes/asm/ghash-neon-armv8.pl
index 4d98bc9a1..0fdcc6296 100644
--- a/crypto/fipsmodule/modes/asm/ghash-neon-armv8.pl
+++ b/crypto/fipsmodule/modes/asm/ghash-neon-armv8.pl
@@ -165,7 +165,7 @@ $code .= <<___;
 .type	gcm_init_neon,%function
 .align	4
 gcm_init_neon:
-	AARCH64_VALID_CALL_TARGET
+	// AARCH64_VALID_CALL_TARGET
 	// This function is adapted from gcm_init_v8. xC2 is t3.
 	ld1	{$t1.2d}, [x1]			// load H
 	movi	$t3.16b, #0xe1
@@ -190,7 +190,7 @@ gcm_init_neon:
 .type	gcm_gmult_neon,%function
 .align	4
 gcm_gmult_neon:
-	AARCH64_VALID_CALL_TARGET
+	// AARCH64_VALID_CALL_TARGET
 	ld1	{$INlo.16b}, [$Xi]		// load Xi
 	ld1	{$Hlo.1d}, [$Htbl], #8		// load twisted H
 	ld1	{$Hhi.1d}, [$Htbl]
@@ -209,7 +209,7 @@ gcm_gmult_neon:
 .type	gcm_ghash_neon,%function
 .align	4
 gcm_ghash_neon:
-	AARCH64_VALID_CALL_TARGET
+	// AARCH64_VALID_CALL_TARGET
 	ld1	{$Xl.16b}, [$Xi]		// load Xi
 	ld1	{$Hlo.1d}, [$Htbl], #8		// load twisted H
 	ld1	{$Hhi.1d}, [$Htbl]
diff --git a/crypto/fipsmodule/modes/asm/ghashv8-armx.pl b/crypto/fipsmodule/modes/asm/ghashv8-armx.pl
index 03cd322ab..caea931c6 100644
--- a/crypto/fipsmodule/modes/asm/ghashv8-armx.pl
+++ b/crypto/fipsmodule/modes/asm/ghashv8-armx.pl
@@ -86,7 +86,7 @@ $code.=<<___;
 .type	gcm_init_clmul,%function
 .align	4
 gcm_init_clmul:
-	AARCH64_VALID_CALL_TARGET
+	// AARCH64_VALID_CALL_TARGET
 	vld1.64		{$t1},[x1]		@ load input H
 	vmov.i8		$xC2,#0xe1
 	vshl.i64	$xC2,$xC2,#57		@ 0xc2.0
@@ -146,7 +146,7 @@ $code.=<<___;
 .type	gcm_gmult_clmul,%function
 .align	4
 gcm_gmult_clmul:
-	AARCH64_VALID_CALL_TARGET
+	// AARCH64_VALID_CALL_TARGET
 	vld1.64		{$t1},[$Xi]		@ load Xi
 	vmov.i8		$xC2,#0xe1
 	vld1.64		{$H-$Hhl},[$Htbl]	@ load twisted H, ...
@@ -200,7 +200,7 @@ $code.=<<___;
 .type	gcm_ghash_clmul,%function
 .align	4
 gcm_ghash_clmul:
-	AARCH64_VALID_CALL_TARGET
+	// AARCH64_VALID_CALL_TARGET
 ___
 $code.=<<___		if ($flavour !~ /64/);
 	vstmdb		sp!,{d8-d15}		@ 32-bit ABI says so
diff --git a/crypto/fipsmodule/sha/asm/sha256-armv4.pl b/crypto/fipsmodule/sha/asm/sha256-armv4.pl
index 63e359ad6..0ef3bc6b7 100644
--- a/crypto/fipsmodule/sha/asm/sha256-armv4.pl
+++ b/crypto/fipsmodule/sha/asm/sha256-armv4.pl
@@ -240,9 +240,9 @@ sha256_block_data_order:
 #ifdef	__APPLE__
 	ldr	r12,[r12]
 #endif
-	tst	r12,#ARMV8_SHA256
+	tst	r12,#(1 << 4)
 	bne	.LARMv8
-	tst	r12,#ARMV7_NEON
+	tst	r12,#(1 << 0)
 	bne	.LNEON
 #endif
 	add	$len,$inp,$len,lsl#6	@ len to point at the end of inp
diff --git a/crypto/fipsmodule/sha/asm/sha512-armv4.pl b/crypto/fipsmodule/sha/asm/sha512-armv4.pl
index df59ef524..d7539a589 100644
--- a/crypto/fipsmodule/sha/asm/sha512-armv4.pl
+++ b/crypto/fipsmodule/sha/asm/sha512-armv4.pl
@@ -302,7 +302,7 @@ sha512_block_data_order:
 #ifdef	__APPLE__
 	ldr	r12,[r12]
 #endif
-	tst	r12,#ARMV7_NEON
+	tst	r12,#(1 << 0)
 	bne	.LNEON
 #endif
 	add	$len,$inp,$len,lsl#7	@ len to point at the end of inp
diff --git a/crypto/fipsmodule/sha/asm/sha512-armv8.pl b/crypto/fipsmodule/sha/asm/sha512-armv8.pl
index a42d96f6f..56e5b820b 100644
--- a/crypto/fipsmodule/sha/asm/sha512-armv8.pl
+++ b/crypto/fipsmodule/sha/asm/sha512-armv8.pl
@@ -186,7 +186,7 @@ $code.=<<___;
 $func:
 ___
 $code.=<<___	if ($SZ==4);
-	AARCH64_VALID_CALL_TARGET
+	// AARCH64_VALID_CALL_TARGET
 #ifndef	__KERNEL__
 #if __has_feature(hwaddress_sanitizer) && __clang_major__ >= 10
 	adrp	x16,:pg_hi21_nc:OPENSSL_armcap_P
@@ -194,12 +194,12 @@ $code.=<<___	if ($SZ==4);
 	adrp	x16,:pg_hi21:OPENSSL_armcap_P
 #endif
 	ldr	w16,[x16,:lo12:OPENSSL_armcap_P]
-	tst	w16,#ARMV8_SHA256
+	tst	w16,#(1 << 4)
 	b.ne	.Lv8_entry
 #endif
 ___
 $code.=<<___;
-	AARCH64_SIGN_LINK_REGISTER
+	// AARCH64_SIGN_LINK_REGISTER
 	stp	x29,x30,[sp,#-128]!
 	add	x29,sp,#0
 
@@ -262,7 +262,7 @@ $code.=<<___;
 	ldp	x25,x26,[x29,#64]
 	ldp	x27,x28,[x29,#80]
 	ldp	x29,x30,[sp],#128
-	AARCH64_VALIDATE_LINK_REGISTER
+	// AARCH64_VALIDATE_LINK_REGISTER
 	ret
 .size	$func,.-$func
 
```

</details>

But it _still_ doesn't get linked into the library so I _still_ get the "unresolved external" errors when building tests 🤯 Did you manage to get past the linking issue somehow?

**Comment by Alovchin91** [comment]
Okay I'm so happy that it builds AND runs that I've opened a draft pull request: #1339

I've also invited you @awakecoding as a collaborator on that fork so if you feel like doing some fixes/additional magic, please feel free 🙂 Branch is `alovchin91/win-arm64`.

Time to sleep now, it's 3am here...

**Comment by Alovchin91** [comment]
Okay, CPU feature detection is done 👍

@briansmith I've left a few comments in the PR #1339 for you to review, but other than that it looks to be... ready? 🤔 (I'm a bit afraid to say so tbh)

**Comment by awakecoding** [comment]
@Alovchin91 there is still a bit of cleanup to be done, it won't build if you don't set $Env:RING_PREGENERATE_ASM='1' once before, and it's a bit annoying because I have to copy nasm.exe inside target/tools/windows/nasm/ every time. I haven't fully figured out how this works, but I believe maybe the assembly files are precompiled by default? The pregeneration does it for all targets, which is why it calls nasm even if I want to target Windows ARM64.

**Comment by Alovchin91** [comment]
@awakecoding Hmm, I don't have that issue but I'm building directly on Surface Pro X, maybe that's why. Did you run `cargo cleanup` beforehand? I'll try it out on an x64 machine tonight.

**Comment by awakecoding** [comment]
I'm currently working on your branch, and I do "git clean -fdx" to get what would happen with a fresh clone. I'd like to improve things a little bit, and I need to understand how it currently works without RING_PREGENERATE_ASM. Unlike you I am cross-compiling from x86_64, so there might be a few differences there.

I have an issue building the tests that seems unrelated to Windows ARM64 because it also affects the x86_64 builds - lots of unresolved imports. Do you have an idea what I might have done wrong? This is using "cargo test --target x86_64-pc-windows-msvc --no-run" but the same happens when targeting aarch64-pc-windows-msvc

![image](https://user-images.githubusercontent.com/295841/128716888-7eaaab7d-d273-4aae-b991-49cf1c2b6f3b.png)

**Comment by Alovchin91** [comment]
Hmm, looks like it cannot reference \*ring\* itself 🤔 Did you `cargo build --target x86_64-pc-windows-msvc` before running `cargo test --target ...`? Also could you please check that on 'main'? I'll be able to check in a couple of hours.

**Comment by awakecoding** [comment]
Ok, I found the source of the problem: I can build and run the tests if I don't explicitly specify the target (x86_64-pc-windows-msvc or aarch64-pc-window-msvc) and use the current default. The thing is that cargo will output files under target/release|debug when using the default target, but will use target/aarch64-pc-window-msvc/release|debug with explicit targets, and that adds an extra directory in the path causing all the breakage. I don't know if there's an easy way around it

**Comment by Alovchin91** [comment]
I wonder what exactly breaks there, since it looks like the CI specifies targets explicitly. Sounds like an issue unrelated to arm64 support to me? 🤔

**Comment by awakecoding** [comment]
Yeah, this is definitely unrelated to ARM64. It's my first time really digging into Ring, and there's a whole bunch of stuff I'm not so certain of.

**Comment by briansmith** [comment]
The current status is documented in BUILDING.md:
* We require (force) clang to be used as the C compiler. Once the next release (0.17.x) is published, clang-cl will also work as the C compiler for published releases. cl.exe won't work yet. This is partially because we need `uint128_t` support in C, but also because BoringSSL only supports clang as the compiler. In order to support cl.exe as the C compiler, we need to verify that there are no ABI differences between cl.exe and clang that affect, in particular, the boundary between C and the assembly. (There shouldn't be many cases where C code is calling assembly functions in *ring* any more.)
* We do dynamic CPU feature detection using the Windows API `IsProcessorFeaturePresent`, just like BoringSSL (see its `cpu-aarch64-win.c`)
* We build the library and the tests in CI, but we don't run the tests, because there's no way to run the tests in CI currently. I have run the tests manually on my Surface Pro X.

**Comment by Alovchin91** [comment]
Great progress!

*ring* was the last thing missing for Rustup on Windows AArch64: https://github.com/rust-lang/rustup/issues/2612 Now we only need to wait for a round of new releases 😅

@briansmith Would you be interested in turning your Surface Pro X collecting dust into a hosted builder for Windows AArch64? I could try it out on my machine and put together some kind of instruction.

_Added bonus: I think I finally decided how to call this flavor of Windows_ 😄

**Comment by briansmith** [comment]
> @briansmith Would you be interested in turning your Surface Pro X collecting dust into a hosted builder for Windows AArch64? I could try it out on my machine and put together some kind of instruction.

Well, it's not "collecting dust" presently as I have it propped up on its built-in stand on my desk.

GitHub recommends against hosting one's own builders in a repo that accepts third-party contributions, for security reasons. My plan is to, for the time being, run the tests on my Surface Pro X when packaging releases and when otherwise I am concerned about breakage.

**Comment by travispaul** [comment]
> Would you be interested in turning your Surface Pro X collecting dust into a hosted builder for Windows AArch64?

Just want to share that the [Snapdragon Dev kits](https://developer.qualcomm.com/hardware/windows-on-snapdragon/snapdragon-developer-kit) are now available, and a much cheaper alternative to a Surface for anyone testing Windows on Arm (while we all wait patiently for Windows aarch64 CI environments to manifest themselves into existence.)

**Comment by Alovchin91** [comment]
Also important news:

https://getwired.com/2022/02/03/can-you-run-windows-on-arm-on-an-apple-silicon-mac-after-all-it-depends/

tl;dr: Though (currently) unsupported, it's perfectly legal to run Windows ARM64 in Parallels on an Apple Silicon Mac.

**Comment by axetroy** [comment]
I saw the above PR, it looked support aarch64 (in the aarch64 Windows)

But can't compile to aarch64 from x86_64?

**Comment by sj6219** [comment]
It was successful in creating a version of arm64(aarch64-pc-windows-msvc)
 using a  program source (https://github.com/briansmith/ring/tree/main) However, it failed to make the x86_64 version. Is this a bug that only happens to me, or has it not been implemented yet?
<img width="867" alt="image" src="https://user-images.githubusercontent.com/3592759/176767868-0f16aed1-408d-4d11-a2ef-c419c3a7fb9c.png">

**Comment by sj6219** [comment]
It seems to be a bug that occurs when building in an environment that uses a non-English language. I changed some files to Unicode and the bug disappeared. 
Modified source is https://github.com/sj6219/ring/tree/unicode

**Comment by awakecoding** [comment]
For other people looking for a temporary solution with a patched 0.16.20 ring dependency, I've written instructions here: https://github.com/briansmith/ring/issues/1514#issuecomment-1258562375

**Comment by briansmith** [comment]
Looking at it now, PR #1554 is a big PR with many non-obvious changes and no explanation to help understand the why/how of the changes. Here I want to explain some of the issues to help others help move things forward.

build.rs has many, many changes between the 0.16 branch and the main branch. It is far from obvious which changes are relevant to the aarch64 port and which changes are 100% backward compatible with what we did in 0.16. One question is whether we should just backport the entire build system (build.rs and related changes) from the main branch to the 0.16 branch wholesale (probably without the symbol prefixing logic), or whether we should try to make only the minimal changes to the 0.16 branch to get aarch64-windows working. I feel like we either need to minimize the diff from the main branch's build.rs, or minimize the diff from the existing 0.16 build.rs.

Presently I am leaning towards just copying the main branch (0.17.0-not-released-yet) build.rs into the 0.16 branch, but removing the symbol prefixing stuff that would require many source code changes outside of build.rs.

We also need to audit which changes to the PerlAsm code in the main branch (mostly inhereted from BoringSSL) are needed to have a working and *safe* aarch64-windows port.

For Windows i686 and x86_64 targets, we ensure that the user doesn't need anything more than Microsoft's build tools and the Rust toolchain installed, when they are using a packaged versoin of *ring* from crates.io. That's why we bundle the generated PerlAsm and even pre-assemble the assembly code into .obj files for i686 and x86_64, so that nasm isn't required. For aarch64, we need to continue with the same design: When using *ring* from crates.io, the user must not need Perl or any third-party tools installed. We *can* require/assume that the user has selected the Aarch64 clang toolchain when installing the MSVC toolchain, so we can rely on clang being there (for aarch64 only). Then we don't need to pre-assembly the aarch64 assembly code into `.o` files. But we do need to still include the PerlAsm output `.S` files into the packaged crate, exactly like we do for other operating systems. It's not clear to me that this is even being done correctly on the main branch. I will investigate this tomorrow as the first step towards making progress here, as it is the trickiest part of the build system.

**Comment by briansmith** [comment]
Another thought: Perhaps it is better to just do the minimum QA to release the main branch as 0.17.0, move all the remaining scheduled 0.17 work to later (possibly 0.18) releases, and update webpki (easy) and Rustls (are they willing) to 0.17.0. Then people who want aarch64-windows support (and many other platforms' support) would "just" need to update to 0.17.

I will investigate this approach more tomorrow.

I invite feedback on these ideas.

**Comment by wmmc88** [comment]
I believe this is the cause of https://github.com/sagiegurari/cargo-make/issues/799. Will ring 0.17.0 be released soon or will the arm64 fixes be backported to 0.16?

**Comment by AArnott** [comment]
#1514 was closed as a dupe of this one, but I wanted to clarify if the issue was the same, since *this* issue doesn't have any mention of the compile errors I see when targeting `aarch64-pc-windows-msvc`. The first (of many) compile errors is:

> .cargo\registry\src\index.crates.io-6f17d22bba15001f\ring-0.16.20\crypto\fipsmodule\bn\internal.h(191): error C2065: 'BN_ULLONG': undeclared identifier

Is this the same issue, or at least fair to group under this one?

**Comment by briansmith** [comment]
Closing this as completed with the *ring* 0.17.0 release. I was able to build and test both natively on ARM64 Windows and cross-compile from x86_64 Windows. PR #1691 clarifies the documentation. I also filed issue #882 for cc-rs to help make this easier so that modifying `%PATH%` will no longer be required in the future.

Thanks for all the help here! I really appreciate it.

**Comment by AArnott** [comment]
Thanks, @briansmith.

I'm struggling to get this to work though. I added this to my Cargo.toml file:

```toml
[patch.crates-io]
ring = "0.17.0"
```

And I get this error:

```
 cargo tree -i ring
    Updating crates.io index
error: failed to resolve patches for `https://github.com/rust-lang/crates.io-index`

Caused by:
  patch for `ring` in `https://github.com/rust-lang/crates.io-index` points to the same source, but patches must point to different sources
```

Any idea why?

**Comment by briansmith** [comment]
Why not just put it in `[dependencies]`?

**Comment by AArnott** [comment]
It isn't a direct dependency. Only an indirect one. Adding it to `[dependencies]` does nothing to alter the indirect ones. Only `[patch.crates-io]` will modify the indirect ones.

This works:
```toml
[patch.crates-io]
ring = { git = "https://github.com/awakecoding/ring", branch = "0.16.20_alpha" }
```

But as it breaks win-x64, I'm eager to adopt your 0.17.0 version that presumably is one version that works for all targets at once.

I also tried a variant of this that consumes _your_ repo. But since you have no 0.17.0 tag, or v0.17 branch, I just pointed it at main:
```toml
[patch.crates-io]
ring = { git = "https://github.com/briansmith/ring", branch = "main" }
```

But that oddly produces this error (where `awakecoding` worked fine):

```
warning: Patch `ring v0.17.0 (https://github.com/briansmith/ring?branch=main#38b9bb7d)` was not used in the crate graph.
Check that the patched package version and available features are compatible
with the dependency requirements. If the patch has a different version from
what is locked in the Cargo.lock file, run `cargo update` to use the new
version. This may also occur with an optional dependency that is not enabled.
```

**Comment by briansmith** [comment]
The patch mechanism doesn't let you patch something using 0.16 with *ring* 0.17. Check the documentation for the patch mechanism to see its version matching requirements.

**Comment by awakecoding** [comment]
Swapping 0.16 with 0.17 won't work even if you fake the version number and cross fingers - all individual dependencies need to be updated to use 0.17, which is why my patch branch just swaps v0.16.20 specifically, which most dependencies use today. To identify what needs updating, use cargo tree like this:

```
sspi-rs> cargo tree -p ring -i
ring v0.16.20
├── rustls v0.21.7
│   ├── hyper-rustls v0.24.1
│   │   └── reqwest v0.11.20
│   │       └── sspi v0.10.1 (C:\wayk\dev\sspi-rs)
│   │           └── sspi-ffi v0.10.1 (C:\wayk\dev\sspi-rs\ffi)
│   ├── reqwest v0.11.20 (*)
│   └── tokio-rustls v0.24.1
│       ├── hyper-rustls v0.24.1 (*)
│       └── reqwest v0.11.20 (*)
├── rustls-webpki v0.101.4
│   └── rustls v0.21.7 (*)
└── sct v0.7.0
    └── rustls v0.21.7 (*)
```

In some other projects, ring gets referenced in a lot more dependencies:

```
devolutions-gateway> cargo tree -p ring -i
ring v0.16.20
├── rustls v0.20.9
│   └── async-rustls v0.3.0
│       └── ngrok v0.13.1
│           └── devolutions-gateway v2023.2.3 (C:\wayk\dev\devolutions-gateway\devolutions-gateway)
│               ├── devolutions-gateway-generators v0.0.0 (C:\wayk\dev\devolutions-gateway\crates\devolutions-gateway-generators)
│               │   [dev-dependencies]
│               │   └── devolutions-gateway v2023.2.3 (C:\wayk\dev\devolutions-gateway\devolutions-gateway) (*)
│               └── generate-openapi v0.0.0 (C:\wayk\dev\devolutions-gateway\tools\generate-openapi)
├── rustls v0.21.7
│   ├── tokio-rustls v0.24.1
│   │   ├── devolutions-gateway v2023.2.3 (C:\wayk\dev\devolutions-gateway\devolutions-gateway) (*)
│   │   └── tokio-tungstenite v0.20.1
│   │       ├── axum v0.6.20
│   │       │   ├── axum-extra v0.8.0
│   │       │   │   └── devolutions-gateway v2023.2.3 (C:\wayk\dev\devolutions-gateway\devolutions-gateway) (*)
│   │       │   └── devolutions-gateway v2023.2.3 (C:\wayk\dev\devolutions-gateway\devolutions-gateway) (*)
│   │       ├── jetsocat v2023.2.3 (C:\wayk\dev\devolutions-gateway\jetsocat)
│   │       └── test-utils v0.0.0 (C:\wayk\dev\devolutions-gateway\crates\test-utils)
│   │           [dev-dependencies]
│   │           ├── jetsocat v2023.2.3 (C:\wayk\dev\devolutions-gateway\jetsocat)
│   │           └── transport v0.0.0 (C:\wayk\dev\devolutions-gateway\crates\transport)
│   │               ├── devolutions-gateway v2023.2.3 (C:\wayk\dev\devolutions-gateway\devolutions-gateway) (*)
│   │               ├── jetsocat v2023.2.3 (C:\wayk\dev\devolutions-gateway\jetsocat)
│   │               └── test-utils v0.0.0 (C:\wayk\dev\devolutions-gateway\crates\test-utils) (*)
│   │   [dev-dependencies]
│   │   └── jet-proto v0.0.0 (C:\wayk\dev\devolutions-gateway\crates\jet-proto)
│   │       └── jetsocat v2023.2.3 (C:\wayk\dev\devolutions-gateway\jetsocat)
│   ├── tokio-tungstenite v0.20.1 (*)
│   ├── tungstenite v0.20.1
│   │   └── tokio-tungstenite v0.20.1 (*)
│   └── ureq v2.7.1
│       [dev-dependencies]
│       └── jet-proto v0.0.0 (C:\wayk\dev\devolutions-gateway\crates\jet-proto) (*)
├── rustls-webpki v0.100.2
│   ├── ureq v2.7.1 (*)
│   └── webpki-roots v0.23.1
│       └── ureq v2.7.1 (*)
├── rustls-webpki v0.101.4
│   └── rustls v0.21.7 (*)
├── sct v0.7.0
│   ├── rustls v0.20.9 (*)
│   └── rustls v0.21.7 (*)
└── webpki v0.22.1
    ├── async-rustls v0.3.0 (*)
    └── rustls v0.20.9 (*)
```

Every single one of those package references needs to be updated to 0.17, otherwise you'll end up with a ring 0.16.20 build that needs patching to build for Windows on ARM. Maybe use my patch branch with a condition in your CI environment if it doesn't build properly for win-x64? It works for [both win-x64 and win-arm64 for me in sspi-rs](https://github.com/Devolutions/sspi-rs/blob/master/.github/workflows/build-native.yml#L119).

**Comment by AArnott** [comment]
> Maybe use my patch branch with a condition in your CI environment if it doesn't build properly for win-x64? 

I've had to use terrible hacks like maintain two cargo.toml files and swap then in based on whether I'm building win-arm64 or any other target. 
IIRC the win-x64 break on the 0.16.20 patch only showed up as missing exported functions, based on which code was actually used by the binary. So it's very conceivable that it worked for you but not for all projects if other projects call functions you don't call.
I'm hoping 0.17.0 will work everywhere.

**Comment by awakecoding** [comment]
> > Maybe use my patch branch with a condition in your CI environment if it doesn't build properly for win-x64?
> 
> I've had to use terrible hacks like maintain two cargo.toml files and swap then in based on whether I'm building win-arm64 or any other target. IIRC the win-x64 break on the 0.16.20 patch only showed up as missing exported functions, based on which code was actually used by the binary. So it's very conceivable that it worked for you but not for all projects if other projects call functions you don't call. I'm hoping 0.17.0 will work everywhere.

indeed, but if you inject the patch line in the CI environment, you can add a condition to only modify the Cargo.toml file for win-arm64, which should at least unblock the issue for now until all dependencies are updated to use 0.17

**Comment by AArnott** [comment]
I need to be able to do this conveniently from my local machine, which depending on the machine is x64 or arm64. So the trick the CI plays has to be easily reproducible and maintainable locally.
I'll keep what I have, I guess while chasing down the dependency trees to get them to upgrade.
